### PR TITLE
🐱‍👤get correct contributor display name from github

### PR DESCRIPTION
**Pull request by OsmondJiang**
#4556 

image

### CODE REVIEW

**Comment by OsmondJiang** [comment]
I will try the graph API to see if it can provide performance

File: src/docfx/lib/github/GitHubAccessor.cs [hunk 1]
```diff
@@ -75,51 +81,103 @@ public async Task<(Error, IEnumerable<GitHubUser>)> GetUsersByCommit(string repo
...
+                }
+
+                var grahpApiResponse = JsonUtility.Deserialize<GithubGraphApiResponse<JObject>>(await response.Content.ReadAsStringAsync(), null);
+                if (grahpApiResponse.Errors != null && grahpApiResponse.Errors.Any())
+                {
+                    return (Errors.GitHubApiFailed(_url, grahpApiResponse.Errors.First().Message), null);
+                }
+
+                var githubUsers = new List<GitHubUser>();
+                if (grahpApiResponse.Data.TryGetValue("repository", out var r) && r is JObject repo &&
```
Comment (yufeih) [inline src/docfx/lib/github/GitHubAccessor.cs]: nit: JsonConvert.DeserializeAnonymousType<>

File: src/docfx/lib/github/GitHubAccessor.cs [hunk 2]
```diff
@@ -75,51 +81,103 @@ public async Task<(Error, IEnumerable<GitHubUser>)> GetUsersByCommit(string repo
...
+
             try
             {
-                var commits = await RetryUtility.Retry(
-                    () => _client.Repository.Commit.GetAll(
-                        repoOwner,
-                        repoName,
-                        new CommitRequest { Sha = commitSha },
-                        new ApiOptions { PageCount = 1, PageSize = 100 }),
+                var response = await RetryUtility.Retry(
```
Comment (yufeih) [inline src/docfx/lib/github/GitHubAccessor.cs]: restrict the change to this method

**Comment by OsmondJiang** [comment]
image

214 calls to get user by login
2370 calls to get user by commits

File: src/docfx/lib/github/GitHubAccessor.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/lib/github/GitHubAccessor.cs]: If the `HttpClientUtility` can't be reused by other component, let's just remove it or keep in inside of `restore`

File: src/docfx/lib/github/GitHubAccessor.cs (hunk 1 above)
Comment (OsmondJiang) [reply src/docfx/lib/github/GitHubAccessor.cs]: great suggestion

File: src/docfx/lib/github/GitHubAccessor.cs [hunk 3]
```diff
@@ -75,51 +83,140 @@ public async Task<(Error, IEnumerable<GitHubUser>)> GetUsersByCommit(string repo
...
+                                        },
+                                    },
+                                },
+                            },
+                        },
+                    },
+                });
+
+                if (grahpApiResponse.errors != null && grahpApiResponse.errors.Any())
+                {
```
Comment (yufeih) [inline src/docfx/lib/github/GitHubAccessor.cs]: Before you are using HttpClient directly, the old `try catch` does not work, so handle these special cases here:
- when rate limit exceed, do not attempt to call GitHub again by setting `_rateLimitError`
- when the repo does not exist or the caller does have access, added it to `_unknownRepos` and do not attempt to call GitHub for this repo again.

File: src/docfx/lib/github/GitHubAccessor.cs [hunk 4]
```diff
@@ -75,51 +83,140 @@ public async Task<(Error, IEnumerable<GitHubUser>)> GetUsersByCommit(string repo
...
-            return new GitHubUser
+            HttpRequestMessage CreateHttpRequest(HttpContent content)
             {
-                Id = commit.Author?.Id,
-                Login = commit.Author?.Login,
-                Name = commit.Commit.Author.Name,
-                Emails = new[] { commit.Commit.Author.Email },
-            };
+                var message = new HttpRequestMessage();
+                message.Headers.Add("User-Agent", "DocFX");
```
Comment (yufeih) [inline src/docfx/lib/github/GitHubAccessor.cs]: prefer `HttpClient.DefaultRequestHeaders` for user agent and authorization, with that, you can simply call `HttpClient.PostAsync` without this helper function.

File: src/docfx/lib/HttpClientUtility.cs [hunk 5]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
+using System.Collections.Generic;
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/HttpClientUtility.cs]: It's use, why i need remove?

File: src/docfx/lib/github/GitHubAccessor.cs [hunk 6]
```diff
@@ -75,51 +83,140 @@ public async Task<(Error, IEnumerable<GitHubUser>)> GetUsersByCommit(string repo
...
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}";
+
+            var request = new
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/github/GitHubAccessor.cs]: I would like to untouch the original method signature

File: src/docfx/lib/github/GitHubAccessor.cs [hunk 7]
```diff
@@ -75,51 +83,140 @@ public async Task<(Error, IEnumerable<GitHubUser>)> GetUsersByCommit(string repo
...
-                return default;
+                if (!response.IsSuccessStatusCode)
+                {
+                    var message = await response.Content.ReadAsStringAsync();
+                    return (Errors.GitHubApiFailed(_url, message), null);
+                }
+
+                var grahpApiResponse = JsonConvert.DeserializeAnonymousType(await response.Content.ReadAsStringAsync(), new
+                {
+                    errors = new[]
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/github/GitHubAccessor.cs]: no, that meets our code requirement and clear for me.

File: src/docfx/lib/github/GitHubAccessor.cs (hunk 4 above)
Comment (OsmondJiang) [reply src/docfx/lib/github/GitHubAccessor.cs]: Only one method is using this http client, and i believe not all are using bearer token, so I would like keep current way.

File: src/docfx/lib/github/GitHubAccessor.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/lib/github/GitHubAccessor.cs]: good point

File: src/docfx/lib/HttpClientUtility.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/HttpClientUtility.cs]: checked the wrong place, removed

**Comment by yufeih** [comment]
With this, can remove `preferExistingName` from github user cache

**Comment by OsmondJiang** [comment]
```
D:\a\1\s\test\docfx.Test\docfx.Test.csproj : warning NU1603: docfx depends on Microsoft.DocAsCode.MarkdigEngine.Extensions (>= 2.42.0-b-190401006-g84b3e9) but Microsoft.DocAsCode.MarkdigEngine.Extensions 2.42.0-b-190401006-g84b3e9 was not found. An approximate best match of Microsoft.DocAsCode.MarkdigEngine.Extensions 2.42.0-b-190407013-gf49407 was resolved.
Error: dotnet test -c Release
At D:\a\1\s\build.ps1:7 char:9
+         throw ("Error: " + $cmd)
```

seems we need upgrade our extension version

cc @928PJY

**Comment by OsmondJiang** [comment]
> With this, can remove `preferExistingName` from github user cache

I would do it in another PR

**Pull request merged**